app/scripts/models/account.js

```diff
@@ -174,9 +177,88 @@ define([
         .then(function () {
           return profileImage;
         });
+    },
+
+    signIn: function (relier) {
+      var self = this;
+      return p().then(function () {
+        var password = self.get('password');
+        var sessionToken = self.get('sessionToken');
+        var email = self.get('email');
+
+        if (password) {
+          return self._fxaClient.signIn(email, password, relier);
+        } else if (sessionToken) {
+          // We have a cached Sync session so just check that it hasn't expired.
+          // The result includes the latest verified state
+          return self._fxaClient.recoveryEmailStatus(sessionToken);
+        } else {
+          throw AuthErrors.toError('UNEXPECTED_ERROR');
+        }
+      })
+      .then(function (updatedSessionData) {
+        self.set(updatedSessionData);
+
+        if (! self.get('verified')) {
+          return self._fxaClient.signUpResend(relier, self.get('sessionToken'));
+        }
+      });
+    },
+
+    signUp: function (relier) {
+      var self = this;
+      return self._fxaClient.signUp(self.get('email'), self.get('password'), relier,
+        {
+          customizeSync: self.get('customizeSync')
+        })
+        .then(function (updatedSessionData) {
+          self.set(updatedSessionData);
+        });
+    },
+
+    saveGrantedPermissions: function (clientId, scope) {
+      var permissions = this.get('grantedPermissions');
+      permissions[clientId] = scopeStrToArray(scope);
+      this.set('grantedPermissions', permissions);
+    },
+
+    hasGrantedPermissions: function (clientId, scope) {
+      if (! scope) {
+        return true;
+      }
+      return this.ungrantedPermissions(clientId, scope).length === 0;
+    },
+
+    ungrantedPermissions: function (clientId, scope) {
+      var granted = this.get('grantedPermissions')[clientId] || [];
+      return detectUngrantedPermissions(scopeStrToArray(scope), granted);
     }
   });
 
+  function set(arr) {
+    var obj = {};
+    for (var i = 0; i < arr.length; i++) {
+      obj[arr[i]] = true;
+    }
+    return Object.keys(obj);
+  }
+
+  function scopeStrToArray(scopes) {
+    return typeof scopes === 'string' ? set(scopes.split(/\s+/g)) : scopes;
+  }
+
+  function detectUngrantedPermissions(requested, granted) {
+    var ungranted = [];
+
+    requested.forEach(function (scope) {
```

Comment:
You might be able to use [_.difference](https://documentcloud.github.io/underscore/#difference) or [_.without](https://documentcloud.github.io/underscore/#without) for this.